src/JSExpr.jl

```diff
@@ -1,207 +1,91 @@
-__precompile__()
-
 module JSExpr
 
-using JSON, MacroTools, WebIO
-export JSString, @js, @js_str, @var, @new
-
-import WebIO: JSString, JSONContext, JSEvalSerialization
-using Observables: AbstractObservable
-
-macro js(expr)
-    :(JSString(string($(jsstring(expr)...))))
-end
-
-# Expressions
+using WebIO: JSString, @js_str
 
-jsexpr(x::JSString) = x.s
-jsexpr(x::Symbol) = (x==:nothing ? "null" : string(x))
-jsexpr(x) = sprint(x) do io, s
-    JSON.show_json(io, JSEvalSerialization(), s)
-end
-jsexpr(x::QuoteNode) = x.value isa Symbol ? jsexpr(string(x.value)) : jsexpr(x.value)
-jsexpr(x::LineNumberNode) = nothing
-
-include("util.jl")
-
-function jsexpr_joined(xs, delim=",")
-    isempty(xs) && return ""
-    F(intersperse(jsexpr.(xs), delim))
-end
+JSString(s::JSString) = s # Definitely move this into WebIO.
+jsstring(xs::JSString...) = JSString(string([x.s for x in xs]...))
+jsstring(xs...) = jsstring(JSString.(xs)...)
 
-jsstring(x) = _simplify(_flatten(jsexpr(x)))
+include("./ast.jl")
 
-function block_expr(args, delim="; ")
-    #print(io, "{")
-    jsexpr_joined(rmlines(args), delim)
-    #print(io, "}")
-end
+"""
+    crawl(expr)
+    crawl(head, ...)
 
-function call_expr(f, args...)
-    if f in [:(=), :+, :-, :*, :/, :%, :(==), :(===),
-             :(!=), :(!==), :(>), :(<), :(<=), :(>=)]
-        return F(["(", jsexpr_joined(args, string(f)), ")"])
-    end
-    F([jsexpr(f), "(", F([jsexpr_joined(args)]), ")"])
-end
+Crawl a given expression and convert it into a `JSNode`.
 
-function obs_get_expr(x)
-    # empty [], special case to get value from an Observable
-    F(["WebIO.getval(", jsexpr(x), ")"])
-end
+There are to versions of `crawl`. The former (`crawl(expr)`) crawls an entire
+expression recursively and converts it into a `JSNode`. The latter form is to
+enable multiple dispatch on expressions using `Val` types.
 
-function obs_set_expr(x, val)
-    # empty [], special case to get value from an Observable
-    F(["WebIO.setval(", jsexpr_joined([x, val]), ")"])
-end
+The expectation is that each dispatched crawl function returns a JSNode by
+calling the crawl-function recursively on deeper expressions.
 
-function jsexpr(o::AbstractObservable)
-    if !haskey(WebIO.observ_id_dict, o)
-        error("No scope associated with observer being interpolated")
+# Examples
+```julia
+crawl(:(foo = "bar"))
+```
+"""
+function crawl(ex::Expr)::JSNode
+    # Recurse into expressions
+    if hasmethod(crawl, typeof((Val(ex.head), ex.args...)))
```

Comment:
```julia
if hasmethod
  ...
else
  ...
end
```
seems a bit strange. Couldn't you define a fallback:
```julia
function crawl(::Val{sym}, args...) where sym
    ex = Expr(sym, args...)
     error("Expression $ex not supported.")
end
```

and just call `crawl(Val(ex.head), ex.args...)` normally? If there is a method it will get it, otherwise you hit the fallback and error.

Reply:
If the `hasmethod` approach turns out to be necessary, rather than `hasmethod(f, typeof(args))` you can just do `applicable(f, args....)`.